### PR TITLE
Use Cava types to prepare for SSZ

### PR DESCRIPTION
**Pull request by atoulme**
## PR Description
* Use Cava Bytes classes.
* Rationalize Hash to use Bytes32 uniformly instead.
* Removed the use of int to use Uint384 throughout instead

### CODE REVIEW

File: util/src/main/java/tech/pegasys/artemis/util/bls/BLSVerify.java [hunk 1]
```diff
@@ -14,14 +14,14 @@
...
 
 import com.google.common.primitives.UnsignedLong;
-import tech.pegasys.artemis.util.bytes.Bytes32;
-import tech.pegasys.artemis.util.uint.UInt384;
+import net.consensys.cava.bytes.Bytes32;
+import net.consensys.cava.bytes.Bytes48;
 
 public class BLSVerify {
 
   // TODO: Need to update UInt384 to UInt384[].
```
Comment (jrhea) [inline util/src/main/java/tech/pegasys/artemis/util/bls/BLSVerify.java]: This comment can be removed.

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java [hunk 2]
```diff
@@ -418,12 +414,12 @@ private boolean validate_proof_of_possession(
    */
   public int process_deposit(
       BeaconState state,
-      int pubkey,
+      Bytes48 pubkey,
       double deposit,
       Bytes32 proof_of_possession,
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java]: proof_of_possession should be 96 bytes.  there is a real and imaginary part to BLS signatures.  There are a couple of way to deal with this.

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java [hunk 3]
```diff
@@ -383,26 +384,21 @@ private int min_empty_validator_index(
    */
   private boolean validate_proof_of_possession(
       BeaconState state,
-      int pubkey,
+      Bytes48 pubkey,
       Bytes32 proof_of_possession,
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java]: proof_of_possession should be 96 bytes.

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java [hunk 4]
```diff
@@ -383,26 +384,21 @@ private int min_empty_validator_index(
...
-            withdrawal_credentials,
-            poc_commitment,
-            randao_commitment,
-            proof_of_possession);
-
-    UInt384 signature =
-        UInt384.valueOf(BytesValue.wrap(proof_of_possession.extractArray()).getInt(0));
+            pubkey, withdrawal_credentials, poc_commitment, randao_commitment, proof_of_possession);
+
+    Bytes48 signature = Bytes48.leftPad(proof_of_possession);
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/state/BeaconState.java]: proof_of_possession should be 96 bytes.

File: util/src/main/java/tech/pegasys/artemis/util/bls/BLSVerify.java [hunk 5]
```diff
@@ -14,14 +14,14 @@
...
   public static boolean bls_verify(
-      UInt384 pubkey, Bytes32 message, UInt384 signature, UnsignedLong domain) {
+      Bytes48 pubkey, Bytes32 message, Bytes48 signature, UnsignedLong domain) {
```
Comment (jrhea) [inline util/src/main/java/tech/pegasys/artemis/util/bls/BLSVerify.java]: signature should be 96 bytes

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java [hunk 6]
```diff
@@ -13,21 +13,21 @@
...
+import net.consensys.cava.bytes.Bytes32;
+import net.consensys.cava.bytes.Bytes48;
 
 public class Attestation {
 
   private AttestationData data;
   private Bytes32 participation_bitfield;
   private Bytes32 custody_bitfield;
-  private UInt384 aggregate_signature;
+  private Bytes48 aggregate_signature;
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java]: aggregate_signature should be 96 bytes or Bytes48[]

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java [hunk 7]
```diff
@@ -13,21 +13,21 @@
...
 
   public Attestation(
       AttestationData data,
       Bytes32 participation_bitfield,
       Bytes32 custody_bitfield,
-      UInt384 aggregate_signature) {
+      Bytes48 aggregate_signature) {
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java]: should be 96 bytes or Bytes48[]

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java [hunk 8]
```diff
@@ -59,11 +59,11 @@ public void setCustody_bitfield(Bytes32 custody_bitfield) {
     this.custody_bitfield = custody_bitfield;
   }
 
-  public UInt384 getAggregate_signature() {
+  public Bytes48 getAggregate_signature() {
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java]: Bytes48[]

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java [hunk 9]
```diff
@@ -59,11 +59,11 @@ public void setCustody_bitfield(Bytes32 custody_bitfield) {
...
     return aggregate_signature;
   }
 
-  public void setAggregate_signature(UInt384 aggregate_signature) {
+  public void setAggregate_signature(Bytes48 aggregate_signature) {
```
Comment (jrhea) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainoperations/Attestation.java]: should be 96 bytes or Bytes48[]

**Review by jrhea — APPROVED** [review]
Updated proof_of_possession and aggregate_signature types to reflect the BLS sig type.  Other than that, lgtm.  I am requesting @schroedingerscode take a look as well.

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/Constants.java [hunk 10]
```diff
@@ -33,7 +33,7 @@
   public static final int MAX_CASPER_VOTES                               = (int) Math.pow(2, 10); // 1,024 votes
   public static final int LATEST_BLOCK_ROOTS_LENGTH                      = (int) Math.pow(2, 13); // 8,192 block roots
   public static final int LATEST_RANDAO_MIXES_LENGTH                     = (int) Math.pow(2, 13); // 8,192 randao mixes
-  public static final UInt384[] EMPTY_SIGNATURE                          = new UInt384[]{UInt384.valueOf(0), UInt384.valueOf(0)};
+  public static final Bytes48[] EMPTY_SIGNATURE                          = new Bytes48[]{Bytes48.ZERO, Bytes48.ZERO};
```
Comment (schroedingerscode) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/Constants.java:36]: This is fine for now, but we will need to decide if we care about just the X coord of the BLS sig (called compressed form) or both the X and Y.

File: ethereum/beaconchain/build.gradle [hunk 11]
```diff
@@ -7,9 +7,11 @@ jar {
 }
 
 dependencies {
-  implementation project(':ethereum:core')
-  implementation project(':ethereum:rlp')
```
Comment (schroedingerscode) [inline ethereum/beaconchain/build.gradle]: Remove newline from top of dependencies block.

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainblocks/ProposalSignedData.java [hunk 12]
```diff
@@ -14,7 +14,7 @@
 package tech.pegasys.artemis.datastructures.beaconchainblocks;
 
 import com.google.common.primitives.UnsignedLong;
-import tech.pegasys.artemis.ethereum.core.Hash;
+import net.consensys.cava.crypto.Hash;
```
Comment (schroedingerscode) [inline ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainblocks/ProposalSignedData.java]: I see in a couple of places that Hash was replaced with Bytes32. We should probably opt for consistency on type usage.

**Review by schroedingerscode — CHANGES_REQUESTED** [review]
Looks good for the most part. Couple comments:
 - ~one whitespace nitpick that I'll fix~ Fixed
 - general question about how Constants.EMPTY_SIGNATURE will look (compressed vs uncompressed) @jrhea thoughts on this?
 - ~found one occurrence where Artemis Hash was replaced with Cava Hash instead of Bytes32 like in most other files.~ Fixed

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainblocks/ProposalSignedData.java (hunk 12 above)
Comment (jrhea) [reply ethereum/beaconchain/src/main/java/tech/pegasys/artemis/datastructures/beaconchainblocks/ProposalSignedData.java]: agreed

File: ethereum/beaconchain/src/main/java/tech/pegasys/artemis/Constants.java (hunk 10 above)
Comment (jrhea) [reply ethereum/beaconchain/src/main/java/tech/pegasys/artemis/Constants.java:36]: good catch on the type.  about the signatures, i opted to follow what the spec called out.  we can revisit the decision on a case by case basis if we decide to go a diff direction

Comment (schroedingerscode) [reply ethereum/beaconchain/src/main/java/tech/pegasys/artemis/Constants.java:36]: Sounds good.

**Review by schroedingerscode — APPROVED** [review]
**Approved.** After the couple of edits it looks good to me.

**Pull request merged**